cranelift/codegen/src/isa/aarch64/lower_inst.rs

```diff
@@ -244,174 +244,79 @@ pub(crate) fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
-                        op: VecRRNarrowOp::Xtn64,
-                        rd: tmp2,
-                        rn: rm,
-                        high_half: false,
-                    });
+            if ty == I128 {
+                let lhs = put_input_in_regs(ctx, inputs[0]);
+                let rhs = put_input_in_regs(ctx, inputs[1]);
+                let dst = get_output_reg(ctx, outputs[0]);
+                assert_eq!(lhs.len(), 2);
+                assert_eq!(rhs.len(), 2);
+                assert_eq!(dst.len(), 2);
 
-                    // Shift the high half components, into the high half.
-                    //   rd = |dg+ch << 32|be+af << 32|
-                    ctx.emit(Inst::VecRRLong {
-                        op: VecRRLongOp::Shll32,
-                        rd,
-                        rn: rd.to_reg(),
-                        high_half: false,
-                    });
+                // 128bit mul formula:
+                //   dst_lo = lhs_lo * rhs_lo
+                //   dst_hi = umulhi(lhs_lo, rhs_lo) + (lhs_lo * rhs_hi) + (lhs_hi * rhs_lo)
+                //
+                // We can convert the above formula into the following
+                // umulh   dst_hi, lhs_lo, rhs_lo
+                // madd    dst_hi, lhs_lo, rhs_hi, dst_hi
+                // madd    dst_hi, lhs_hi, rhs_lo, dst_hi
+                // mul     dst_lo, lhs_lo, rhs_lo
 
-                    // Multiply the low components together, and accumulate with the high
-                    // half.
-                    //   rd = |rd[1] + cg|rd[0] + ae|
-                    ctx.emit(Inst::VecRRR {
-                        alu_op: VecALUOp::Umlal,
-                        rd,
-                        rn: tmp2.to_reg(),
-                        rm: tmp1.to_reg(),
-                        size: VectorSize::Size32x2,
-                    });
+                ctx.emit(Inst::AluRRR {
+                    alu_op: ALUOp::UMulH,
+                    rd: dst.regs()[1],
+                    rn: lhs.regs()[0],
+                    rm: rhs.regs()[0],
+                });
+                ctx.emit(Inst::AluRRRR {
+                    alu_op: ALUOp3::MAdd64,
+                    rd: dst.regs()[1],
+                    rn: lhs.regs()[0],
+                    rm: rhs.regs()[1],
+                    ra: dst.regs()[1].to_reg(),
+                });
+                ctx.emit(Inst::AluRRRR {
+                    alu_op: ALUOp3::MAdd64,
+                    rd: dst.regs()[1],
+                    rn: lhs.regs()[1],
+                    rm: rhs.regs()[0],
+                    ra: dst.regs()[1].to_reg(),
+                });
+                ctx.emit(Inst::AluRRRR {
+                    alu_op: ALUOp3::MAdd64,
+                    rd: dst.regs()[0],
+                    rn: lhs.regs()[0],
+                    rm: rhs.regs()[0],
+                    ra: zero_reg(),
+                });
+            } else if ty.is_vector() {
+                for ext_op in &[
```

Comment:
One possible way to structure this is to perhaps check for `is_vector` and i64x2 early on at the top of this block with early-returns if they match, and if that fails all the remaining cases share the logic of 

```rust
                    let rn = put_input_in_reg(ctx, inputs[0], NarrowValueMode::None);
                    let rm = put_input_in_reg(ctx, inputs[1], NarrowValueMode::None);
                    let rd = get_output_reg(ctx, outputs[0]).only_reg().unwrap();
```

as a prefix.

Just a thought though, happy to defer to you who work on this much more than I!

Reply:
So even though I don't like the duplication, or how much conditional stuff is going on here, we would only be able to factor out one collection of these calls as the I128 case uses, ever so slightly, different calls to handle multiple registers.   So I think I prefer the clarity.